otherlibs/systhreads/st_stubs.c

```diff
@@ -277,18 +277,17 @@ static caml_thread_t caml_thread_new_info(void)
 
 static value caml_thread_new_descriptor(value clos)
 {
-  value mu = Val_unit;
+  CAMLparam1(clos);
+  CAMLlocal1(mu);
   value descr;
-  Begin_roots2 (clos, mu)
-    /* Create and initialize the termination semaphore */
-    mu = caml_threadstatus_new();
-    /* Create a descriptor for the new thread */
-    descr = caml_alloc_small(3, 0);
-    Ident(descr) = Val_long(atomic_fetch_add(&thread_next_id, +1));
-    Start_closure(descr) = clos;
-    Terminated(descr) = mu;
-  End_roots();
-  return descr;
+  /* Create and initialize the termination semaphore */
+  mu = caml_threadstatus_new();
+  /* Create a descriptor for the new thread */
+  descr = caml_alloc_small(3, 0);
+  Ident(descr) = Val_long(atomic_fetch_add(&thread_next_id, +1));
+  Start_closure(descr) = clos;
+  Terminated(descr) = mu;
```

Comment:
While we're modernising: these macros and `struct caml_thread_descr` are I think a hangover from older times. I think `struct caml_thread_descr` can be deleted, and the three macros redefined as:

```c
#define Ident(v) Field(v, 0)
#define Start_closure(v) Field(v, 1)
#define Terminated(v) FIeld(v, 2)
```
which could then allow:
```suggestion
  descr = caml_alloc_3(0,
    Val_long(atomic_fetch_add(&thread_next_id, +1)),
    clos,
    mu);
```

Reply:
Thanks, fixed.